File: basic_filling_algorithm/test_algorithm.py
# Заполняет 2 область, ограниченную 1
# Добавить затравочный пиксель в стек;
# 3 образно помечен старт
def print_map(local_map):
    for i in range(len(local_map)):
        for j in range(len(local_map[i])):
            print(local_map[i][j], end = ' ')
        print()

def main():
    '''
    local_map = [[0,1,1,1,1,0,0,1,0,0],
                 [1,0,0,0,0,1,1,3,1,0],
                 [1,0,0,0,0,1,0,0,0,1],
                 [0,1,0,0,0,0,0,0,0,1],
                 [0,0,1,0,0,0,0,0,1,0],
                 [0,0,0,1,1,1,1,1,0,0]]
    '''
    local_map = [[3,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0],
                 [0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0],
                 [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
                 [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
                 [0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0],
                 [0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0],
                 [0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0],
                 [0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0],
                 [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]]

    # map_stack = [[1,7]]
    map_stack = [[0,0]]
    print_map(local_map)
    while(len(map_stack)):
        # Извлечение пикселя из стека
        robot_pos = map_stack[0]; map_stack.pop(0)
        
        x_min = robot_pos[0]
        while((x_min > 0) and (local_map[x_min - 1][robot_pos[1]] != 1)):
            x_min -= 1
            
        x_max = robot_pos[0]
        while((x_max < len(local_map) - 1) and (local_map[x_max + 1][robot_pos[1]] != 1)):
            x_max += 1

        # Закрасить отрезок (X_min,Y)--(X_max,Y) цветом A;
        for x in range(x_min, x_max + 1):
            local_map[x][robot_pos[1]] = 2
            
        # Обработка строки сверху
        flag_color = True
        for x in range(x_min, x_max + 1):
            if (local_map[x][robot_pos[1] - 1] != 1 and local_map[x][robot_pos[1] - 1] == 0):
                if (flag_color):
                    map_stack.append([x, robot_pos[1] - 1])
                    flag_color = False
            else:
                flag_color = True
                
        # Обработка строки снизу
        flag_color = True
        for x in range(x_min, x_max + 1):
            if (local_map[x][robot_pos[1] + 1] != 1 and local_map[x][robot_pos[1] + 1] == 0):
                if (flag_color):
                    map_stack.append([x, robot_pos[1] + 1])
                    flag_color = False
            else:
                flag_color = True
                
        # print()
        # print_map(local_map)
        # print()

    print()
    print_map(local_map)
    print()

if __name__ == '__main__':
    main()

File: basic_filling_algorithm/test_algorithm.cpp
#include <stdio.h>

/*void pop(int ***array)
{
	//int **tmp_array = *array;
	int **a = *array;

	//int tmp_array[6][10];
	//int *b = (int*) tmp_array;
	//b[0] = a[1];
	a[0] = a[1];
	for (int i = 1; i < 5; ++i)
	{
		//for (int j = 0; j < 10; )
		a[i] = a[i + 1];
	}

	//*array = tmp_array;
}*/
void print_list(int local_map[6][10])
{
	for (int i = 0; i < 6; ++i)
	{
		for (int j = 0; j < 10; ++j)
			printf("%d ", local_map[i][j]);
	}
	printf("\n");
}
void main(void)
{
	int local_map[6][10] =
	{
		{ 0, 1, 1, 1, 0, 0, 1, 0, 0, 1 },
		{ 1, 0, 0, 0, 1, 1, 3, 1, 1, 0 },
		{ 1, 0, 0, 0, 0, 1, 0, 0, 0, 1 },
		{ 0, 1, 0, 0, 1, 0, 0, 0, 0, 1 },
		{ 0, 0, 1, 0, 0, 0, 0, 0, 1, 0 },
		{ 0, 0, 0, 1, 1, 1, 1, 1, 0, 0 }
	};
	int map_stack[60][2];
	map_stack[0][0] = 2; map_stack[0][1] = 6;
	int size = 1;

	while (map_stack[0][0] != 0 && map_stack[0][1] != 0)
	{
		int robot_pos[2];
		robot_pos[0] = map_stack[0][0]; robot_pos[1] = map_stack[0][1];
		// pop
		map_stack[0][0] = map_stack[1][0]; map_stack[0][1] = map_stack[1][1];
		for (int i = 1; i < 59; ++i)
		{
			map_stack[i][0] = map_stack[i + 1][0]; map_stack[i][1] = map_stack[i + 1][1];
		}
		map_stack[59][0] = 0; map_stack[59][1] = 0;

		int x_min = robot_pos[0];
		while ((x_min > 0) && (local_map[x_min - 1][robot_pos[1]] != 1))
			x_min--;

		int x_max = robot_pos[0];
		while ((x_max < 5) && (local_map[x_max - 1][robot_pos[1]] != 1))
			x_max++;

		for (int x = x_min; x <= x_max; ++x)
		{
			local_map[x][robot_pos[1]] = 2;
		}

		int flag_color = 1;
		for (int x = x_min; x <= x_max; ++x)
		{
			if (local_map[x][robot_pos[1] - 1] != 1 && local_map[x][robot_pos[1] - 1] == 0)
			{
				if (flag_color)
				{
					map_stack[size][0] = x; map_stack[size][1] = robot_pos[1] - 1;
					size++;
					flag_color = 0;
					//forward();
					//delay(1000);
					//stop();
				}
			}
			else
			{
				flag_color = 1;
			}
		}

		flag_color = 1;
		for (int x = x_min; x <= x_max; ++x)
		{
			if (local_map[x][robot_pos[1] + 1] != 1 && local_map[x][robot_pos[1] + 1] == 0)
			{
				if (flag_color)
				{
					map_stack[size][0] = x; map_stack[size][1] = robot_pos[1] + 1;
					size++;
					flag_color = 0;
					//forward();
					//delay(1000);
					//stop();
				}
			}
			else
			{
				flag_color = 1;
			}
		}
		print_list(local_map);
	}
}